### PR TITLE
Asyncio Task identifier support

### PR DESCRIPTION
**Pull request by br0wn9**
Used to support wsgi server developed using asyncio library

For example: https://github.com/br0wn9/lin

### CODE REVIEW

File: src/werkzeug/local.py [hunk 1]
```diff
@@ -5,11 +5,22 @@
...
-# thread ident.
+# thread ident. Asyncio Task ident is preferred.
+
 try:
-    from greenlet import getcurrent as get_ident
-except ImportError:
-    from threading import get_ident
+    from asyncio import get_running_loop, current_task
+
+    get_running_loop()
```
Comment (davidism) [inline src/werkzeug/local.py]: Why is this call needed? What if there's not an event loop at import time, but there is in the context where the local is being used?

What if this does succeed, but the application isn't using asyncio, they're using threads or gevent?

File: src/werkzeug/local.py [hunk 2]
```diff
@@ -5,11 +5,22 @@
 
 # Each thread has its own greenlet, use that as the identifier for the
 # context. If greenlets are not available fall back to the current
-# thread ident.
+# thread ident. Asyncio Task ident is preferred.
```
Comment (davidism) [inline src/werkzeug/local.py]: Asyncio is now "preferred", but is only given a short sentence at the *end* of this explanation comment.

File: src/werkzeug/local.py [hunk 3]
```diff
@@ -5,11 +5,22 @@
...
-    from threading import get_ident
+    from asyncio import get_running_loop, current_task
+
+    get_running_loop()
+
+    def get_ident():
+        return id(current_task())
+
+
+except (ImportError, RuntimeError):
```
Comment (davidism) [inline src/werkzeug/local.py]: What would cause an `ImportError`? A `RuntimeError`?

**Comment by davidism** [comment]
Since I'm not an expert on asyncio, I'm going to need a *ton* of explanation about changes related to it. I'm particularly concerned that it assumes asyncio is in use if it's importable, even if the application is using threads or gevent still.

Looking at asgiref's local implementation, there seems to be a lot more to account for than just the current task. https://github.com/django/asgiref/blob/master/asgiref/local.py Is that relevant here?

**Comment by davidism** [comment]
What about contextvars? Can we just say asyncio support requires 3.7+ and use those? https://docs.python.org/3/library/contextvars.html

**Comment by davidism** [comment]
`asgiref.local` just had a problem with assuming the id of the task remained unique: https://github.com/django/asgiref/commit/e4367c571fd1833bda0b7ff0bd68d924916352da

**Comment by pgjones** [comment]
I used to do something similar in Quart, but I switched to [ContextVars](https://gitlab.com/pgjones/quart/-/blob/master/src/quart/local.py) (as they also work with Trio/Curio). I'd suggest using those instead. (This was my plan).

**Comment by davidism** [comment]
Sorry to drop all these comments at once, I do appreciate that you worked on this, it's just very important that we get it right.

**Comment by br0wn9** [comment]
The get_running_loop function is just testing whether it's running in the asyncio task context,

If you do not asyncio task context raise RuntimeError

**Comment by davidism** [comment]
That doesn't seem correct. What is it's not asyncio when it's imported, but it is one there's a request being handled? What if there's an event loop but it's not being used?

**Comment by br0wn9** [comment]
> That doesn't seem correct. What if it's not asyncio when it's imported, but it is one there's a request being handled? What if there's an event loop but it's not being used?

If there is no event loop, or the event loop is not used. The get_running_loop function throws RuntimeError. 

```
    from asyncio import get_running_loop, current_task  # asyncio is not supported, raise ImportError

    get_running_loop()  # no event loop or event loop not used, raise RuntimeError
   
    def get_ident():
        return id(current_task())
```

**Comment by davidism** [comment]
I'm going through and doing some setup, then I do a theoretical `run_asgi` to start the event loop and the development server. But `werkzeug.local` was imported during setup, there was no loop running yet, and so all locals use the current thread id because `get_running_loop` failed.

Putting that aside for a moment. As @pgjones pointed out, they already tried this approach and discarded it in favor of contextvars. And there's some other issues that haven't been addressed as well.

**Comment by br0wn9** [comment]
> I'm going through and doing some setup, then I do a theoretical `run_asgi` to start the event loop and the development server. But `werkzeug.local` was imported during setup, there was no loop running yet, and so all locals use the current thread id because `get_running_loop` failed.
> 
> Putting that aside for a moment. As @pgjones pointed out, they already tried this approach and discarded it in favor of contextvars. And there's some other issues that haven't been addressed as well.

The wsgi application must be loaded after the event loop starts, There is something wrong with the scene you mentioned

The pr I submitted was not to support asgi, This is to make asyncio available to wsgi applications

You can actually test it.  [lin](https://github.com/br0wn9/lin)

**Comment by pgjones** [comment]
@br0wn9 could you give your thoughts on #1778? It is a contextvar alternative to this.

**Comment by davidism** [comment]
closing in favor of #1778

**Pull request closed without merging**